profiler.js

```diff
@@ -168,80 +168,160 @@ module.exports = {
       "resolver"
     ]);
 
-    var paths = this.convert_to_absolute_paths(options.paths, options.base_path);
+    var resolver = options.resolver;
 
-    function findRequiredSources(dependsGraph, done) {
-      var required = {};
+    // Fetch the whole contract set
+    find_contracts(options.contracts_directory, (err, allPaths) => {
+      if(err) return callback(err);
 
-      function hasBeenTraversed(import_path) {
-        return required[import_path] != null;
-      }
+      // Solidity test files might have been injected. Include them in the known set.
+      options.paths.forEach(_path => {
+        if (!allPaths.includes(_path)) {
+          allPaths.push(_path)
+        }
+      });
 
-      function include(import_path) {
-        //console.log("Including: " + file)
+      var updates = self.convert_to_absolute_paths(options.paths, options.base_path).sort();
+      var allPaths = self.convert_to_absolute_paths(allPaths, options.base_path).sort();
 
-        required[import_path] = dependsGraph.node(import_path);
-      }
+      var allSources = {};
+      var compilationTargets = {};
 
-      function walk_down(import_path) {
-        if (hasBeenTraversed(import_path)) {
-          return;
-        }
+      // Get all the source code
+      self.resolveAllSources(resolver, allPaths, (err, resolved) => {
+        if(err) return callback(err);
 
-        include(import_path);
+        // Generate hash of all sources including external packages - passed to solc inputs.
+        var resolvedPaths = Object.keys(resolved);
+        resolvedPaths.forEach(file => allSources[file] = resolved[file].body)
 
-        var dependencies = dependsGraph.successors(import_path);
+        // Exit w/out minimizing if we've been asked to compile everything, or nothing.
+        if (self.listsEqual(options.paths, allPaths)){
+          return callback(null, allSources, {});
+        } else if (!options.paths.length){
+          return callback(null, {}, {});
+        }
 
-        // console.log("At: " + import_path);
-        // console.log("   Dependencies: ", dependencies);
+        // Seed compilationTargets with known updates
+        updates.forEach(update => compilationTargets[update] = resolved[update].body);
 
-        if (dependencies.length > 0) {
-          dependencies.forEach(walk_down);
-        }
-      }
+        // While updates: dequeue
+        async.whilst(() => updates.length > 0, updateFinished => {
+          var currentUpdate = updates.shift();
+          var files = allPaths.slice();
 
-      function walk_from(import_path) {
-        if (hasBeenTraversed(import_path)) {
-          return;
-        }
+          // While files: dequeue and inspect their imports
+          async.whilst(() => files.length > 0, fileFinished => {
```

Comment:
Since you removed the explicit graph, might be worth adding a comment at the start of this traversal, to document the general process of queueing and dequeuing as a strategy (ie why we are doing this in the first place)